pkg/virt-api/rest/subresource.go

```diff
@@ -1259,3 +1263,154 @@ func (app *SubresourceAPIApp) VMIAddVolumeRequestHandler(request *restful.Reques
 func (app *SubresourceAPIApp) VMIRemoveVolumeRequestHandler(request *restful.Request, response *restful.Response) {
 	app.removeVolumeRequestHandler(request, response, true)
 }
+
+func (app *SubresourceAPIApp) SEVFetchCertChainRequestHandler(request *restful.Request, response *restful.Response) {
+	if !app.clusterConfig.WorkloadEncryptionSEVEnabled() {
+		writeError(errors.NewBadRequest(fmt.Sprintf(featureGateDisabledFmt, virtconfig.WorkloadEncryptionSEV)), response)
+		return
+	}
+
+	validate := func(vmi *v1.VirtualMachineInstance) *errors.StatusError {
+		if !vmi.IsScheduled() && !vmi.IsRunning() {
```

Comment:
> Can you explain why !vmi.IsScheduled() part is essential?

In order to fetch the certificates we only need to know the correct worker node. The VMI in that case does not have to be in the `Running` state. It can still be `Scheduled`. This is in fact the case while it waits for the SEV session parameters fom the user.

> The error seems somewhat misleading that way

Yeah, agree. I will adjust it to reflect better the situation here. Thanks for pointing that.

Reply:
@vasiliy-ul I think you need to remove the `!vmi.IsScheduled` right? If you use libvirt API you need to be able to talk to virt-launcher

Reply:
No, this is a bit tricky. The VMI cannot transition to the `Running` state without the session parameters being provided. And a user can construct those only after fetching the SEV platform certs. So the flow is like the following:
- A VMI requesting SEV attestation is launched with empty `Session` and `DHCert` in yaml
- It gets **scheduled** to the proper node with SEV support
- `virt-handler` waits for the session parameters and meanwhile prevents the VMI from transitioning to the `Running` state
- User fetches the platform certs from the node
- User builds session parameters and calls `sev/setupsession` API
- The VMI spec gets patched with the session parameters
- `virt-handler` notices the session parameters and allows the transition to the `Running` state
- Then libvirt starts the domain in `Paused` state.

Reply:
ok, but then you need to verify that the virt-launcher pod is running, otherwise you cannot talk to virt-launcher process, right?

Reply:
I don't know it depends it can fail to fetch the platform certificates because virt-launcher hasn't been started yet. For example, I'm not sure if the VMI could be result is scheduled but virt-launcher hasn't been started yet because we are pulling a container image or because one of the init containers takes long

Reply:
Looked through the code a bit. So here the `VMI` transitions to the `Scheduled` phase:

https://github.com/kubevirt/kubevirt/blob/a3aaf1ace40c44c9dafaf2c091851d5cdb829fd7/pkg/virt-controller/watch/vmi.go#L535

This code is inside the condition:

https://github.com/kubevirt/kubevirt/blob/a3aaf1ace40c44c9dafaf2c091851d5cdb829fd7/pkg/virt-controller/watch/vmi.go#L509

The function `IsPodReady` checks if the `virt-launcher` pod and the **compute container** are already running:

https://github.com/kubevirt/kubevirt/blob/a3aaf1ace40c44c9dafaf2c091851d5cdb829fd7/pkg/virt-controller/watch/vmi.go#L893-L915

Unless I fail to see something obvious that should work for the fetch-certs case. Anyway, if for whatever reason the pod is not ready, it will just return an error and the fetching can be retried.